### PR TITLE
[ITSASU-1998] Do not allow editing of tax year if not permitted.

### PR DESCRIPTION
**Pull request by aug24**

### CODE REVIEW

File: app/views/agent/AgentTaskList.scala.html [hunk 1]
```diff
@@ -69,18 +69,26 @@
...
     <span class="app-task-list__task-name">
       {if(viewModel.taxYearSelectedAndConfirmed) {
-      <a href={controllers.agent.routes.TaxYearCheckYourAnswersController.show(editMode=true).url} aria-describedby="section_1_thing_1_status" id="select_tax_year_check_your_answer">
-          {viewModel.taxYearSelection.map(_.accountingYear) match {
-              case Some(Current) => messages(s"${messagePrefix}.content.section1.current-tax-year", (accountingPeriodService.currentTaxYear - 1).toString, accountingPeriodService.currentTaxYear.toString)
-              case Some(Next) => messages(s"${messagePrefix}.content.section1.next-tax-year", accountingPeriodService.currentTaxYear.toString, (accountingPeriodService.currentTaxYear + 1).toString)
-              case None => messages(s"${messagePrefix}.content.section1.action")
-          }}
-      </a>
+        if(!viewModel.permitTaxYearChange)
```
Comment (sb-dev) [inline app/views/agent/AgentTaskList.scala.html:85]: `taxYearSelectedAndConfirmed` is going to return `false` so `taxYearText` needs to be moved somewhere else.

File: app/models/common/TaskListModel.scala [hunk 2]
```diff
@@ -24,9 +24,10 @@ case class TaskListModel(taxYearSelection: Option[AccountingYearModel],
                          selfEmployments: Seq[SelfEmploymentData],
                          selfEmploymentAccountingMethod: Option[AccountingMethod],
                          ukProperty:  Option[PropertyModel],
-                         overseasProperty: Option[OverseasPropertyModel]) {
+                         overseasProperty: Option[OverseasPropertyModel],
+                         permitTaxYearChange: Boolean = true) {
 
-  val taxYearSelectedAndConfirmed: Boolean = taxYearSelection.exists(_.confirmed)
+  val taxYearSelectedAndConfirmed: Boolean = taxYearSelection.exists(_.confirmed) || (taxYearSelection.nonEmpty && !permitTaxYearChange)
```
Comment (sb-dev) [inline app/models/common/TaskListModel.scala:30]: `taxYearSelection` is going to be empty so the `taxYearSelection.nonEmpty` check is not necessary.

Comment (aug24) [reply app/models/common/TaskListModel.scala:30]: I don't understand.  If `taxYearSelection` is empty then that would never be true.
What this does is consider the result "confirmed" which is usually a user action, if the tax year is actually being mandated.

File: app/views/agent/AgentTaskList.scala.html (hunk 1 above)
Comment (aug24) [reply app/views/agent/AgentTaskList.scala.html:85]: Why is it?  It should return true if it has a value and the user is not permitted to change it.  That's precisely what the other comment is about.

**Comment by aug24** [comment]
I realised that the editable flag could be moved into the tax year model.  That cleaned it up quite a bit.

File: app/models/common/TaskListModel.scala [hunk 3]
```diff
@@ -18,40 +18,37 @@ package models.common
...
 
 case class TaskListModel(taxYearSelection: Option[AccountingYearModel],
                          selfEmployments: Seq[SelfEmploymentData],
                          selfEmploymentAccountingMethod: Option[AccountingMethod],
-                         ukProperty:  Option[PropertyModel],
+                         ukProperty: Option[PropertyModel],
                          overseasProperty: Option[OverseasPropertyModel]) {
 
-  val taxYearSelectedAndConfirmed: Boolean = taxYearSelection.exists(_.confirmed)
+  val taxYearSelectedAndConfirmed: Boolean = taxYearSelection.exists(a => !a.editable || a.confirmed)
```
Comment (sb-dev) [inline app/models/common/TaskListModel.scala:28]:
```suggestion
  val taxYearSelectedAndConfirmed: Boolean = taxYearSelection.exists(taxYear => !taxYear.editable || taxYear.confirmed)
```

File: app/models/common/TaskListModel.scala [hunk 4]
```diff
@@ -18,40 +18,37 @@ package models.common
...
 
-  val taxYearSelectedNotConfirmed: Boolean = taxYearSelection.exists(!_.confirmed)
+  val taxYearSelectedNotConfirmed: Boolean = taxYearSelection.exists(a => a.editable && !a.confirmed)
```
Comment (sb-dev) [inline app/models/common/TaskListModel.scala:30]:
```suggestion
  val taxYearSelectedNotConfirmed: Boolean = taxYearSelection.exists(taxYear => taxYear.editable && !taxYear.confirmed)
```

File: app/models/common/TaskListModel.scala [hunk 5]
```diff
@@ -18,40 +18,37 @@ package models.common
...
                          overseasProperty: Option[OverseasPropertyModel]) {
 
-  val taxYearSelectedAndConfirmed: Boolean = taxYearSelection.exists(_.confirmed)
+  val taxYearSelectedAndConfirmed: Boolean = taxYearSelection.exists(taxYear => !taxYear.editable || taxYear.confirmed)
 
-  val taxYearSelectedNotConfirmed: Boolean = taxYearSelection.exists(!_.confirmed)
+  val taxYearSelectedNotConfirmed: Boolean = taxYearSelection.exists(taxYear => taxYear.editable && !taxYear.confirmed)
 
-  val ukPropertyComplete: Boolean =  ukProperty.exists(_.confirmed)
+  val permitTaxYearChange: Boolean = taxYearSelection.exists(a => a.editable)
```
Comment (sb-dev) [inline app/models/common/TaskListModel.scala:32]:
```suggestion
  val permitTaxYearChange: Boolean = taxYearSelection.exists(_.editable)
```

**Review by sb-dev — APPROVED** [review]

**Pull request merged**